File: 3802-largest-local-values-in-a-matrix-ii/largest-local-values-in-a-matrix-ii.py
class Solution:
    def countLocalMaximums(self, mat: List[List[int]]) -> int:
        n=len(mat)
        m=len(mat[0])

        pos=[[] for _ in range(201)]
        for i in range(n):
            for j in range(m):
                    pos[mat[i][j]].append((i, j))

        res=0
        for i in range(n):
            for j in range(m):
                x=mat[i][j]
                if x == 0:
                    continue

                flag=True
                for val in range(x+1, 201):
                    if not flag:
                        break
                    for r, c in pos[val]:
                        dir=abs(r-i)
                        dic=abs(c-j)

                        if dir<=x and dic<=x:
                            if dir == x and dic == x:
                                continue;
                            flag=False;
                            break
                    
                if flag:
                    res+=1
            
        return res

File: Medium
Medium/3802-largest-local-values-in-a-matrix-ii/largest-local-values-in-a-matrix-ii.java
class Solution{
    public int countLocalMaximums(int[][] mat){
        int n=mat.length;
        int m=mat[0].length;

        ArrayList<int[]>[] pos=new ArrayList[201];
        for(int i=0; i<=200; i++){
            pos[i]=new ArrayList<>();
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                // if(mat[i][j] != 0)
                    pos[mat[i][j]].add(new int[]{i, j});
            }
        }

        int res=0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                int x=mat[i][j];
                if(x == 0)
                    continue;

                boolean flag=true;
                for(int val=x+1; val<=200 && flag; val++){
                    for(int[] p: pos[val]){
                        int r=p[0];
                        int c=p[1];
                        int dir=Math.abs(r-i);
                        int dic=Math.abs(c-j);

                        if(dir<=x && dic<=x){
                            if(dir == x && dic == x)
                                continue;
                            flag=false;
                            break;
                        }
                    }
                }
                if(flag)
                    res++;
            }
        }

        return res;
    }
}